Add find() support within our DFU traverser

dongahn opened the pull request:
This PR introduces `find()` support within our DFU traverser
and integrates it with `resource-query` for testing.

`find()` traverses the resource graph, evaluate each visiting
vertex and emit those resources whose states are matched
with the criteria passed in.

- To evaluate each resource vertex, we view
    each resource vertex as a set of predicates:
   ` { predicate1, predicate2, ..., predicateN }`.
   Example predicates are "is status down?" (`status=down`),
   "is scheduled state allocated?" (`sched-now=allocated`)
    or "is future scheduled state reserved?" (`sched-future=reserved`).

- Add and use the `expr_eval_target_vtx_t` adapter class
   on each visiting resource vertex so that we can evaluate
   each queried predicate on its current state without having
   to modify the resource vertex classes themselves.

- Because we allow the partial marking of a vertex,
   the state of an ancestor vertex can often determine
   the effective state of a child vertex.
   For instance, if the status of a compute node vertex
   is "down", the status of a core underneath it must also be
   "down" even if the  actually marked status is still "up".
    To support this case, we add and use the
    `vtx_predicates_override_t` class to aid in passing
    ancestor predicates down to the offspring resources
    throughout the traversal.

- `find()` takes in a base writer object and match criteria.
    Performs the DFU visit on the resource graph and conducts
    the following on certain visit events:
    1) On each preorder visit, it builds the predicates
    of the visiting vertex and passes it to  the next vertex visit
    (depth first visit on the dominant subsystem and up visit on
    auxiliary subsystems).
    2) On each postorder visit, it builds `expr_eval_vtx_target_t`
    based on the state of the visiting vertex *and* the predicates set
    passed down from the parent and evaluates the
    expression on this mixed state.

Integrate `find()` with `resource-query` and
adds a test using its `find expression` subcommand.

Add a fix to `resource-query` for `set|get-status`
subcommands to avoid a race between resource-query's
prompt outputs and the testing output by supporting
`-t` option within  `set|get-status`. 
Adjust `t3024-resource-status.t` with respect to that.

dongahn commented:
PR #684 depends on this PR. 

This PR also depends on one commit from PR #667, it would be best to merge PR #667 first.

dongahn commented:
CI isn't happy; I added a few changes. Once Trivis CI turns green, this should be ready for your reviews.

dongahn commented:
Travis is green now!

dongahn commented:
@Mergifyio rebase

dongahn commented:
@Mergefy.io rebase

dongahn commented:
@Mergifyio rebase

milroy commented on resource/traversers/dfu_impl.hpp:
```diff
@@ -205,6 +209,31 @@ class dfu_impl_t {
+     *                   Currently supported expressions are
+     *                   "status={up|down}", "sched-now={allocated|free}",
+     *                   "sched-future={reserved|free}, or any combination
+     *                   of them separated with "and", "or", or a whitespace
+     *                   which is also interpreted as "and" logical
+     *                   operator of two expressions. Parentheses
```
It looks like this answers my previous question about the whitespace in `resource/evaluators/test/expr_eval_test01.cpp`.  Can you add a similar explanation there?

dongahn replied:
Sure. Will do.

milroy commented on t/t3025-resource-find.t:
```diff
@@ -0,0 +1,256 @@
+    diff up.filt.json full.filt.json
+'
+
+test_expect_success 'mark down 2 cores and find status=down' '
```
```suggestion
test_expect_success 'mark down 3 cores and find status=down' '
```

dongahn replied:
Oops. Thanks.

milroy commented on resource/utilities/command.cpp:
```diff
@@ -478,9 +478,12 @@ int cmd_set_status (std::shared_ptr<resource_context_t> &ctx,
         ctx->db->metadata.by_path.find (vtx_path);
```
It's great that you found this race condition!

milroy commented:
This contribution is really great!

Most of my suggestions and comments are on the order of nits.  The main exception is related to parsing parentheses.  My manual tests mostly behave as expected, but I did find some strange results related to an unbalanced terminating parenthesis.  Using the tiny example cluster for example:

```bash
resource-query> set-status /tiny0/rack0/node0/socket0/core0 down
resource-query> find (status=down) and (sched-now=free)
      ---------------core0[1:x]
      ------------socket0[1:x]
      ---------node0[1:x]
      ------rack0[1:x]
      ---tiny0[1:x]
INFO: =============================
INFO: EXPRESSION="(status=down) and (sched-now=free)"
INFO: =============================
resource-query> find (status=down) and (sched-now=free))
      ---------------core0[1:x]
      ------------socket0[1:x]
      ---------node0[1:x]
      ------rack0[1:x]
      ---tiny0[1:x]
INFO: =============================
INFO: EXPRESSION="(status=down) and (sched-now=free))"
INFO: =============================
resource-query> find (status=down) and (sched-now=free)))
ERROR: find: invalid criteria: (status=down) and (sched-now=free))).
```

dongahn commented:
> I did find some strange results related to an unbalanced terminating parenthesis. Using the tiny example cluster for example:

Thanks @milroy. I will go through all of your comments. But it feels like this one is some "one-off" bug somewhere. I will look!

milroy commented:
> But it feels like this one is some "one-off" bug somewhere.

Yes, that's the feeling I get since my other tests are successful.

dongahn commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -0,0 +1,352 @@
+/*****************************************************************************\
```
A nest parenthesized expression should be evaluated before the outer expression is evaluated.

milroy replied:
What I meant to ask was whether this PR evaluates nested parenthetic expressions via recursion (which certainly didn't come across in my comment).  That's why I listed the call sequence.  This is a more of a curiosity and not necessary for the merge.

dongahn replied:
Ah... sorry about that. Yes, it is handled through recursion.

dongahn commented on resource/evaluators/expr_eval_target.hpp:
```diff
@@ -0,0 +1,76 @@
+class expr_eval_target_base_t {
+public:
+
+    /*! Validate if a predicate expression is valid.
```
Thanks. Good catch.

dongahn commented on resource/evaluators/expr_eval_vtx_target.hpp:
```diff
@@ -0,0 +1,123 @@
+    bool is_initialized () const;
+
+private:
+    bool m_initialized{false};
```
I actually have seen no leading whitespace version of uniform initialization more often. (But then I have seen no leading whitespace version of function calls more often :-)

Since we don't have a coding standard for this and I think I have used no leading whitespace version previously. We probably should keep it this way for now.

dongahn commented on resource/evaluators/test/expr_eval_test01.cpp:
```diff
@@ -0,0 +1,205 @@
+\*****************************************************************************/
+
+#if HAVE_CONFIG_H
+# include <config.h>
```
Great catch!

dongahn commented on resource/evaluators/test/expr_eval_test01.cpp:
```diff
@@ -0,0 +1,205 @@
+void build_simple_expr (std::vector<std::string> &expr_vector)
+{
+    expr_vector.push_back ("  status=up      ");
+    expr_vector.push_back ("  status=up          sched-now=allocated    	");
```
Yes this is intended. I wanted to test extra whitespace cases and also no operator between predicates that are assumed to be "and". (modeled from Unix `find` tool).

dongahn commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -0,0 +1,352 @@
+/*****************************************************************************\
```
Better! Thanks.

dongahn commented on resource/traversers/dfu_impl.cpp:
```diff
@@ -575,6 +575,71 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
+    rc = nchildren + 1;
+done:
+    m_trav_level--;
+    return rc;
```
The higher number of subtree matches, the higher rc will be. It is just indicate whether you need to select the visiting vertex event if it doesn't actually match with the expression. To form a graph, you need to select that vertex in that case. It is not necessarily more desirable than `rc == 1`.

If you are asking why we don't limit `rc` to be 0, -1 or 1, I don't have a good answer. `rc > 1` seems more natural to propagate richer information to the upper function frame in this recursion. So I chose it this way.

dongahn commented on resource/evaluators/test/expr_eval_test01.cpp:
```diff
@@ -0,0 +1,205 @@
+
+void build_simple_expr (std::vector<std::string> &expr_vector)
+{
+    expr_vector.push_back ("  status=up      ");
+    expr_vector.push_back ("  status=up          sched-now=allocated    	");
+    expr_vector.push_back ("status=up and sched-now=allocated");
+    expr_vector.push_back ("status=up or sched-now=allocated");
+    expr_vector.push_back ("status=up sched-now=free");
+    expr_vector.push_back ("status=up   and sched-now=free");
+    expr_vector.push_back ("status=up or sched-now=free");
+    expr_vector.push_back ("status=down   sched-now=allocated");
+    expr_vector.push_back ("status=down and sched-now=allocated");
+    expr_vector.push_back ("status=down or sched-now=allocated");
+    expr_vector.push_back ("status=down sched-now=free");
+    expr_vector.push_back ("status=down and sched-now=free");
+    expr_vector.push_back ("status=down or sched-now=free");
+    expr_vector.push_back ("sched-now=allocated status=up");
+    expr_vector.push_back ("sched-now=allocated and status=up");
+    expr_vector.push_back ("sched-now=allocated or status=up");
+    expr_vector.push_back ("sched-now=free or status=up");
+    expr_vector.push_back ("sched-now=free or   status=up");
+    expr_vector.push_back ("sched-now=free or status=up");
+    expr_vector.push_back ("sched-now=allocated or status=down");
+    expr_vector.push_back ("sched-now=allocated and status=down");
+    expr_vector.push_back ("sched-now=allocated or status=down");
+    expr_vector.push_back ("sched-now=free status=down");
+    expr_vector.push_back ("sched-now=free and status=down");
+    expr_vector.push_back ("sched-now=free  or          status=down");
```
This is intended: test to see if extra whitespaces are ignored.

dongahn commented:
@Mergifyio rebase

dongahn commented:
@milroy: Thank you for your review agains. I addressed all of your comments and added them as fixup commits. The only change that has already been squashed is for your commit message change request for the predicate. 

> But it feels like this one is some "one-off" bug somewhere. I will look!

As suspected, this was indeed due to an one-off bug where I should have use `<=` instead of `<`. 

Great finding!

SteVwonder commented on t/t3025-resource-find.t:
```diff
@@ -0,0 +1,256 @@
+	set-status /tiny0/rack0/node0/socket1 up
+	set-status /tiny0/rack0/node0/socket1/core19 down
+	set-status /tiny0/rack0/node0/socket0/core16 down
+	find (status=up and sched-now=allocated) or sched-future=reserved
```
Just to clarify my own understanding, is it possible for a node to be down and still have a new job reservation made on it?

dongahn replied:
Yes it is possible under `resource-query` and our resource graph infrastructure. But how we use this in `sched-fluxion-resource`, this will not be possible as "down" marking will call sched_loop which will clear up all reservations and then rebuild. While rebuilding the alloc/reservations, the down resources won't be used.

SteVwonder commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -289,7 +289,7 @@ int expr_eval_api_t::validate (const std::string &e,
     at = next;
 
-    while (at < e.find_last_not_of (" \t")) {
+    while (at <= e.find_last_not_of (" \t")) {
```
I didn't see the original comment from Dan that lead to this fixup commit, but is there a test (either unit or integration) that covers this edge case now?  Not sure how easy this would be to test for.

dongahn replied:
I just realized this. I will add some to our unit tests.

dongahn commented on t/t3025-resource-find.t:
```diff
@@ -0,0 +1,256 @@
+    ${query} -L ${grugs} -F rv1_nosched -S CA -P high -t up.out < cmds002 &&
+    cat up.out | grep -v "INFO" > up.json &&
+    remove_times up.json > up.filt.json &&
+    diff up.filt.json full.filt.json
```
Oh nice. I didn't know. Will use `test_cmp`

dongahn commented on t/t3025-resource-find.t:
```diff
@@ -0,0 +1,256 @@
+
+. $(dirname $0)/sharness.sh
+
+sock1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test001.yaml"
+c1g1m1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test002.yaml"
+node1_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test003.yaml"
+full_job="${SHARNESS_TEST_SRCDIR}/data/resource/jobspecs/find/test004.yaml"
```
No objection. I was just following the convention so far. I do agree a descriptive name is better.

dongahn commented on resource/traversers/dfu_impl.cpp:
```diff
@@ -575,6 +575,71 @@ int dfu_impl_t::dom_dfv (const jobmeta_t &meta, vtx_t u,
+    bool down = (*m_graph)[u].status == resource_pool_t::status_t::DOWN;
+    bool allocated = !(*m_graph)[u].schedule.allocations.empty ();
+    bool reserved = !(*m_graph)[u].schedule.reservations.empty ();
+    Flux::resource_model::vtx_predicates_override_t p_overriden = p;
```
Yes. `p_overriden` is first copy-constructed from `p` and then updated as passed with the following line `set()`. This way each method frame will have its own copy of `p_overriden` without having to leak it to the caller frame (since p is passed by reference.). I may be missing something though.

dongahn replied:
I will mark it resolved for now. Let me know if I missed something.

SteVwonder replied:
It might be worth throwing a comment above this line calling out the copy-construction for those less familiar with C++ (like me).  I thought this was just a simple name change, I didn't realize there was copying going on.  Thanks for the clarification.

dongahn commented on resource/evaluators/expr_eval_vtx_target.cpp:
```diff
@@ -0,0 +1,117 @@
+        goto done;
+    }
+    if (p == "status")
+        rc = (x == "down" || x == "DOWN" || x == "up" || x == "UP")? 0 : -1;
+    else if (p == "sched-now")
+        rc = (x == "allocated"
+             || x == "ALLOCATED" || x == "free" || x == "FREE")? 0 : -1;
+    else if (p == "sched-future")
+        rc = (x == "reserved"
+             || x == "RESERVED" || x == "free" || x == "FREE")? 0 : -1;
```
Heh. Yeah I thought about it and got lazy. C++ standard library must have an easy to do this. Will add that support.

dongahn replied:
`std::transform` does the magic.

```diff
@@ -41,18 +43,19 @@ int expr_eval_vtx_target_t::validate (const std::string &p,
                                       const std::string &x) const
 {
     int rc = -1;
+    std::string lcx = x;
+
     if (!m_initialized) {
         errno = EINVAL;
         goto done;
     }
+    std::transform (x.begin(), x.end(), lcx.begin(), ::tolower);
     if (p == "status")
-        rc = (x == "down" || x == "DOWN" || x == "up" || x == "UP")? 0 : -1;
+        rc = (lcx == "down" || lcx == "up")? 0 : -1;
     else if (p == "sched-now")
```

dongahn commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -0,0 +1,352 @@
+
+    m_level++;
+
+    if ( (delim = e.find_first_of ("=")) == std::string::npos) {
```
The intent was to say `p=x` key-value pair encodes `p(x)` in predicate calculus. We probably don't want a predicate actually to be expressed as p(x) since the double use of parenthesis, using `p=x` to encode p(x) would be okay? or is it too confusing?

SteVwonder replied:
Ah! I think the thing I was missing was that that is how it is represented in predicate calculus.  I totally agree we don't want the extra parenthesis, lest this become a lisp variant ;)

dongahn commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -0,0 +1,352 @@
+ *                                                                          *
+ ****************************************************************************/
+
+int expr_eval_api_t::validate (const std::string &e,
```
Yeah, this is possible and good idea. I think we can add `expr_eval_err_t` in a similar way to jansson (was it `json_err_t`?) and support you are asking for . For the interest of time, though, we may create a ticket for this and address it as part of a future small enhancement ticket?

dongahn replied:
Open Issue #699.

dongahn commented on resource/evaluators/expr_eval_api.hpp:
```diff
@@ -0,0 +1,112 @@
+                        size_t at, size_t &next, bool &result);
+    int evaluate_pred (pred_op_t op, bool result2, bool &result1) const;
+
+    int m_level = 0;
```
The intent was to add some nice indented debug message using that. But yes since we don't use this in this PR, I will remove the use of this member variable.

dongahn commented on resource/evaluators/expr_eval_api.cpp:
```diff
@@ -0,0 +1,352 @@
+    }
+    if ( (xws = e.find_first_of (" \t", nws)) == std::string::npos)
+        xws = e.length ();
+    op = pred_op_t::AND;
+    next = at;
+    if (e.substr (nws, xws - nws) == "and") {
+        next = xws;
```
I haven't test it but `X=foo pr Y=bar` should be interpreted as `X=foo and pr and Y=bar` and by the time you get to `pr`, the evaluation system will invalidate the expression. I will test.

dongahn replied:
Just confirmed it is the case:

```console
./resource-query -L../../t/data/resource/grugs/tiny.graphml
INFO: Loading a matcher: CA
resource-query> find status=up pr sched-now=allocated
ERROR: find: invalid criteria: status=up pr sched-now=allocated.
```

dongahn commented on resource/evaluators/test/expr_eval_test01.cpp:
```diff
@@ -0,0 +1,205 @@
+int expr_eval_test_target_t::evaluate (const std::string &p,
+                                       const std::string &x, bool &result) const
+{
+    int rc = -1;
+    if ( (rc = validate (p, x)) < 0)
+        return rc;
+    result = true;
+    return rc;
```
Hmm. The intent is for this function always to return the same return code as `validate`. And when it returns 0, `result` is always set to be true for testing purpose.

dongahn commented on resource/evaluators/test/expr_eval_test01.cpp:
```diff
@@ -0,0 +1,205 @@
+        "and (sched-now=free and) status=down)");
+    expr_vector.push_back ("(status=up (sched-now=allocated) "
+        "and (sched-now=free)status=down)");
+    expr_vector.push_back ("(status=up)(sched-now=allocated)"
+        "(sched-now=free)(status=down)");
```
Yes, you need either "and" or "or" or one or more whitespaces which is implicit "and". -- Modeled after Unix `find` command.

dongahn commented on resource/evaluators/expr_eval_vtx_target.cpp:
```diff
@@ -0,0 +1,117 @@
+        goto done;
+    if (p == "status") {
+        if (x == "down" || x == "DOWN") {
+            result = m_overriden.status_down
```
Great catch!

dongahn commented:
@SteVwonder: I believe I have addressed all the issues. Could you take a one last look at the fixup commits (since this is pretty late at night I'm not that capable of looking at those with confidence :-)

Also I opened Issue at https://github.com/flux-framework/flux-sched/issues/699 per your request at https://github.com/flux-framework/flux-sched/pull/693#discussion_r457792401

SteVwonder commented:
Looks like the valgrind failure was just intermittent.  All good for a squash and merge :)

dongahn commented:
Thanks @SteVwonder and @milroy. I squashed all fixup commits and forced a push. When CI turns green, this will go in.
